File: androidApp/src/main/kotlin/dev/lildua/oddly/ui/state/OddlyAppState.kt
package dev.lildua.oddly.ui.state

import androidx.compose.runtime.Composable
import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import dev.lildua.oddly.data.seed.ChallengeSeed
import dev.lildua.oddly.data.seed.SampleData
import dev.lildua.oddly.domain.model.AppSettings
import dev.lildua.oddly.domain.model.Category
import dev.lildua.oddly.domain.model.Challenge
import dev.lildua.oddly.domain.model.ChallengeCompletion
import dev.lildua.oddly.domain.model.StreakInfo
import dev.lildua.oddly.domain.model.UserProfile
import dev.lildua.oddly.domain.usecase.ChallengeSelector
import dev.lildua.oddly.domain.usecase.DayActivity
import dev.lildua.oddly.domain.usecase.StreakCalculator
import kotlinx.datetime.LocalDate
import kotlinx.datetime.LocalTime
import kotlin.random.Random

/**
 * In-memory application state, seeded from [SampleData].
 *
 * This stands in for the repository + ViewModel layer that arrives in Phase 2.
 * Everything is real, derived state — completing a challenge genuinely updates
 * the streak, calendar, statistics and XP — so the screens can be reviewed with
 * live behaviour rather than static mock values.
 */
@Stable
class OddlyAppState(
    val today: LocalDate,
) {
    private val _completions = mutableStateListOf<ChallengeCompletion>().apply {
        addAll(SampleData.completions)
    }

    val completions: List<ChallengeCompletion> get() = _completions

    var profile: UserProfile by mutableStateOf(SampleData.profile)
        private set

    var settings: AppSettings by mutableStateOf(SampleData.settings)

    var interests: Set<Category> by mutableStateOf(SampleData.profile.interests)

    /** The challenge shown on Home; stable for the whole day (spec §6.1). */
    var todayChallenge: Challenge by mutableStateOf(SampleData.todayChallenge)
        private set

    /** Set when the user opens a challenge from anywhere other than Home. */
    var skippedCount: Int by mutableStateOf(SampleData.SKIPPED_COUNT)
        private set

    /** XP and Humanity gained by the most recent completion, for the reward screen. */
    var lastReward: Reward? by mutableStateOf(null)
        private set

    private val random = Random(SampleData.today.toEpochDays())

    val streak: StreakInfo
        get() = StreakCalculator.calculate(_completions, today)

    val weekActivity: List<DayActivity>
        get() = StreakCalculator.recentActivity(_completions, today)

    val completedToday: Boolean
        get() = _completions.any { it.date == today }

    val totalCompleted: Int get() = _completions.size

    val completionRatePercent: Int
        get() {
            val total = _completions.size + skippedCount
            return if (total == 0) 0 else (_completions.size * 100) / total
        }

    val exploredCategoryCount: Int
        get() = _completions
            .mapNotNull { ChallengeSeed.byId(it.challengeId)?.category }
            .distinct()
            .size

    fun completionsOn(date: LocalDate): List<ChallengeCompletion> =
        _completions.filter { it.date == date }.sortedByDescending { it.completedAt }

    fun challengeOf(completion: ChallengeCompletion): Challenge? =
        ChallengeSeed.byId(completion.challengeId)

    fun isCompleted(challengeId: String): Boolean =
        _completions.any { it.challengeId == challengeId }

    fun isCompletedToday(challengeId: String): Boolean =
        _completions.any { it.date == today && it.challengeId == challengeId }

    /** Record a completion and award XP, levelling up if the bar fills. */
    fun complete(challenge: Challenge, at: LocalTime = LocalTime(9, 41)) {
        if (_completions.any { it.date == today && it.challengeId == challenge.id }) return

        _completions.add(
            ChallengeCompletion(
                id = "completion_${challenge.id}_${today.toEpochDays()}",
                challengeId = challenge.id,
                date = today,
                completedAt = at,
                xpEarned = challenge.rewardXp,
                humanityPercent = challenge.humanityPercent,
            ),
        )

        var level = profile.level
        var xp = profile.xpInLevel + challenge.rewardXp
        while (xp >= UserProfile.xpForLevel(level)) {
            xp -= UserProfile.xpForLevel(level)
            level += 1
        }
        val leveledUp = level > profile.level
        profile = profile.copy(level = level, xpInLevel = xp)

        lastReward = Reward(
            xp = challenge.rewardXp,
            humanityPercent = challenge.humanityPercent,
            leveledUp = leveledUp,
            newLevel = level,
        )
    }

    /** Swap today's challenge for another one (spec §2.3). */
    fun reroll(category: Category? = null) {
        val next = ChallengeSelector.select(
            library = ChallengeSeed.all,
            interests = interests,
            completions = _completions,
            level = profile.level,
            today = today,
            restrictTo = category,
            excludeChallengeId = todayChallenge.id,
            random = random,
        )
        if (next != null) {
            skippedCount += 1
            todayChallenge = next
        }
    }

    /** Make [challenge] the active one, e.g. after picking from the library. */
    fun chooseTodayChallenge(challenge: Challenge) {
        todayChallenge = challenge
    }

    fun toggleInterest(category: Category) {
        interests = if (category in interests) interests - category else interests + category
    }

    fun clearReward() {
        lastReward = null
    }

    /** Wipes local data (spec §S15). Used by the Reset action in Settings. */
    fun resetAllData() {
        _completions.clear()
        profile = profile.copy(level = 1, xpInLevel = 0)
        skippedCount = 0
        lastReward = null
    }

    data class Reward(
        val xp: Int,
        val humanityPercent: Int,
        val leveledUp: Boolean,
        val newLevel: Int,
    )
}

@Composable
fun rememberOddlyAppState(): OddlyAppState =
    remember { OddlyAppState(today = SampleData.today) }

File: iosApp/iosApp/State/OddlyAppState.swift
import Observation
import SharedLogic

/// In-memory application state, seeded from `SampleData`.
///
/// This stands in for the repository layer that arrives in Phase 2. Everything
/// is real, derived state — completing a challenge genuinely updates the streak,
/// calendar, statistics and XP — so the screens can be reviewed with live
/// behaviour rather than static mock values.
///
/// The Android app has the same class backed by Compose snapshot state; this is
/// the SwiftUI half, and the two stay deliberately parallel so the shared
/// domain layer remains the only place business rules live.
@Observable
final class OddlyAppState {

    struct Reward {
        let xp: Int
        let humanityPercent: Int
        let leveledUp: Bool
        let newLevel: Int
    }

    let today: LocalDate

    private(set) var completions: [ChallengeCompletion]
    private(set) var profile: UserProfile
    var settings: AppSettings
    var interests: Set<ChallengeCategory>

    /// The challenge shown on Home; stable for the whole day (spec §6.1).
    private(set) var todayChallenge: Challenge

    /// How many challenges were started and set aside — drives completion rate.
    private(set) var skippedCount: Int

    /// XP and Humanity gained by the most recent completion, for the reward screen.
    private(set) var lastReward: Reward?

    init() {
        let sample = SampleData.shared
        today = sample.today
        completions = sample.completions
        profile = sample.profile
        settings = sample.settings
        interests = sample.profile.interests
        todayChallenge = sample.todayChallenge
        skippedCount = Int(sample.SKIPPED_COUNT)
    }

    // MARK: - Derived state

    var streak: StreakInfo {
        StreakCalculator.shared.calculate(completions: completions, today: today)
    }

    var weekActivity: [DayActivity] {
        StreakCalculator.shared.recentActivity(completions: completions, today: today, days: 7)
    }

    var completedToday: Bool {
        completions.contains { $0.date == today }
    }

    var totalCompleted: Int { completions.count }

    var completionRatePercent: Int {
        let total = completions.count + skippedCount
        return total == 0 ? 0 : (completions.count * 100) / total
    }

    var exploredCategoryCount: Int {
        Set(completions.compactMap { ChallengeSeed.shared.byId(id: $0.challengeId)?.category }).count
    }

    var quoteOfTheDay: Quote {
        QuoteSeed.shared.forDayIndex(dayIndex: today.toEpochDays())
    }

    // MARK: - Lookups

    func completionsOn(_ date: LocalDate) -> [ChallengeCompletion] {
        completions
            .filter { $0.date == date }
            .sorted { $0.completedAt.toSecondOfDay() > $1.completedAt.toSecondOfDay() }
    }

    func challenge(of completion: ChallengeCompletion) -> Challenge? {
        ChallengeSeed.shared.byId(id: completion.challengeId)
    }

    /// Resolves completions to their challenges, dropping any whose challenge is
    /// no longer in the library.
    func resolved(_ source: [ChallengeCompletion]) -> [CompletionRef] {
        source.compactMap { completion in
            guard let challenge = challenge(of: completion) else { return nil }
            return CompletionRef(id: completion.id, completion: completion, challenge: challenge)
        }
    }

    func isCompleted(_ challengeId: String) -> Bool {
        completions.contains { $0.challengeId == challengeId }
    }

    func isCompletedToday(_ challengeId: String) -> Bool {
        completions.contains { $0.date == today && $0.challengeId == challengeId }
    }

    // MARK: - Mutations

    /// Record a completion and award XP, levelling up if the bar fills.
    func complete(_ challenge: Challenge, at time: LocalTime = LocalTime(hour: 9, minute: 41, second: 0, nanosecond: 0)) {
        guard !isCompletedToday(challenge.id) else { return }

        completions.append(
            ChallengeCompletion(
                id: "completion_\(challenge.id)_\(today.epochDays)",
                challengeId: challenge.id,
                date: today,
                completedAt: time,
                xpEarned: challenge.rewardXp,
                humanityPercent: challenge.humanityPercent,
                note: nil
            )
        )

        var level = profile.level
        var xp = profile.xpInLevel + challenge.rewardXp
        while xp >= UserProfile.companion.xpForLevel(level: level) {
            xp -= UserProfile.companion.xpForLevel(level: level)
            level += 1
        }
        let leveledUp = level > profile.level
        profile = profile.with(level: level, xpInLevel: xp)

        lastReward = Reward(
            xp: Int(challenge.rewardXp),
            humanityPercent: Int(challenge.humanityPercent),
            leveledUp: leveledUp,
            newLevel: Int(level)
        )
    }

    /// Swap today's challenge for another one (spec §2.3).
    @discardableResult
    func reroll(category: ChallengeCategory? = nil) -> Challenge? {
        let next = ChallengeSelector.shared.select(
            library: ChallengeSeed.shared.all,
            interests: interests,
            completions: completions,
            level: profile.level,
            today: today,
            restrictTo: category,
            excludeChallengeId: todayChallenge.id,
            random: KotlinRandom.Default.shared
        )
        guard let next else { return nil }
        skippedCount += 1
        todayChallenge = next
        return next
    }

    /// Make [challenge] the active one, e.g. after picking from the library.
    func chooseTodayChallenge(_ challenge: Challenge) {
        todayChallenge = challenge
    }

    func toggleInterest(_ category: ChallengeCategory) {
        if interests.contains(category) {
            interests.remove(category)
        } else {
            interests.insert(category)
        }
    }

    func clearReward() {
        lastReward = nil
    }

    /// Wipes local data (spec §S15). Used by the Reset action in Settings.
    func resetAllData() {
        completions.removeAll()
        profile = profile.with(level: 1, xpInLevel: 0)
        skippedCount = 0
        lastReward = nil
    }
}
